ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/MicrometerMetrics.kt

```diff
@@ -0,0 +1,185 @@
+import io.ktor.request.*
+import io.ktor.routing.*
+import io.ktor.util.*
+import io.ktor.util.pipeline.*
+import io.micrometer.core.instrument.*
+import io.micrometer.core.instrument.binder.*
+import io.micrometer.core.instrument.binder.jvm.*
+import io.micrometer.core.instrument.binder.system.*
+import io.micrometer.core.instrument.config.*
+import io.micrometer.core.instrument.distribution.*
+import java.util.concurrent.atomic.*
+
+
+/**
+ * Enables Micrometer support when installed. Exposes the following metrics:
+ * <ul>
+ *     <li><code>ktor.http.server.requests.active</code>: Gauge - The amount of active ktor requests</li>
+ *     <li><code>ktor.http.server.requests</code>: Timer - Timer for all requests. By default no percentiles or
+ *       histogram is exposed. Use the [Configuration.timerCustomizer] to enable these.
+ *       Tags:
+ *       <ul>
+ *           <li><code>local</code>: The host and port of the request uri (e.g. 'www.ktor.io')</li>
+ *           <li><code>method</code>: The http method (e.g. 'GET')</li>
+ *           <li><code>route</code>: The use ktor route used for this request. (e.g. '/some/path/{someParameter}')
+ *           <li><code>status</code>: The http status code or the Exception name thrown during the request (e.g. '200', 'java.lang.IllegalArgumentException')
+ *        <ul>
+ *     <li>
+ *  <ul>
+ */
+class MicrometerMetrics(
+    private val registry: MeterRegistry,
+    timerDistributionConfig: DistributionStatisticConfig,
+    private val tagCustomizer: MicrometerMetrics.TagBuilder.(ApplicationCall) -> Unit
+) {
+
+    private val active = registry.gauge("$baseName.requests.active", AtomicInteger(0))
+
+    init {
+        enableTimerDistributionConfig(timerDistributionConfig)
+    }
+
+    private fun enableTimerDistributionConfig(timerDistributionConfig: DistributionStatisticConfig) {
+        registry.config().meterFilter(object : MeterFilter {
+            override fun configure(id: Meter.Id, config: DistributionStatisticConfig): DistributionStatisticConfig =
+                if (id.name == requestTimerName)
+                    timerDistributionConfig.merge(config)
+                else
+                    config
+        })
+    }
+
+    /**
+     * Configures this Feature
+     * @property registry The meter registry where the meters are registered. Mandatory
+     * @property meterBinders The binders that are automatically bound to the registry. Default: [ClassLoaderMetrics],
+     * [JvmMemoryMetrics], [ProcessorMetrics], [JvmGcMetrics], [ProcessorMetrics], [JvmThreadMetrics], [FileDescriptorMetrics]
+     * @property distributionStatisticConfig configures the histogram and/or percentiles for all request timers. By
+     * default 50%, 90% , 95% and 99% percentiles are configured
+     * @property tagBuilder is called to customize the tags of the timer for each call. by default this is a no-op
+     */
+    class Configuration {
+
+        lateinit var registry: MeterRegistry
```

Comment:
The reason why I did this initially was to allow to keep a reference to the meter's instance in DI, especially without a `getRegistry()` I suggested before, it would be impossible to declare the scrape endpoint...

Do you have a better suggestion?

Reply:
The user must have the chance to specify the `MeterRegistry` they want to have. @dave08 and me might prefer the `PrometheusMeterRegistry` others might prefer the Influx, Elastic, ... Implementation. Each implementation needs different config params, so we must provide the user a possibility to inject it from outside. 

```suggestion
        var registry: MeterRegistry? = null
```
and then make a nullcheck when initializing the feature

Reply:
Yes, and it should be a clean error message, something like `"No micrometer registry specified. Configure in the feature configuration section."`

Reply:
Rather than defaulting to null, I suggest defaulting to empty `CompositeMeterRegistry`, if you don't want to use the global registry by default. Interaction with an empty composite is a no-op. Then the type can be non-null.

Reply:
The question would be if we tell the developer "Hey you missed to configure something" of if we provide him with a valid no-op default and then the user needs to find out why we don't provide him any metrics. I think the later case would be worse in user experience then my inital approach with lateinit, where the user would have had at least an NPE guiding him to the problem.

Reply:
I event found a better solution. You can check if a `lateinit` field is initialized. So there is no need to make it nullable.